Marginally improve pytest collection for top-level test files

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53682 Add meta device to generic device testing framework, skip NotImplementedError
* #53681 Make storage access throw for meta tensors
* **#53617 Marginally improve pytest collection for top-level test files**
* #53611 Fix inaccurate dispatch table for fill_
* #53610 Convert a few more checks to raise NotImplementedError

I'm trying to make `pytest test/*.py` work--right now, it fails during
test collection.  This removes a few of the easier to fix pytest
collection problems one way or another.  I have two remaining problems
which is that the default dtype is trashed on entry to test_torch.py and
test_cuda.py, I'll try to fix those in a follow up.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26918377](https://our.internmc.facebook.com/intern/diff/D26918377)

ezyang committed:
Marginally improve pytest collection for top-level test files

I'm trying to make `pytest test/*.py` work--right now, it fails during
test collection.  This removes a few of the easier to fix pytest
collection problems one way or another.  I have two remaining problems
which is that the default dtype is trashed on entry to test_torch.py and
test_cuda.py, I'll try to fix those in a follow up.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Marginally improve pytest collection for top-level test files"


I'm trying to make `pytest test/*.py` work--right now, it fails during
test collection.  This removes a few of the easier to fix pytest
collection problems one way or another.  I have two remaining problems
which is that the default dtype is trashed on entry to test_torch.py and
test_cuda.py, I'll try to fix those in a follow up.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

mruberry commented on test/test_overrides.py:
```diff
@@ -315,9 +315,16 @@ def generate_tensor_like_torch_implementations():
+    # the problem.  A more proper fix is to make the "not tested" check
+    # a test on its own, and to make sure the monkeypatch is only installed
+    # for the span of the relevant test (and deleted afterwards)
+    testing_ignore = {"sample_functional"}
```
Why not address this by removing the monkey patch, then?

ezyang replied:
Because I don't really understand what the test is doing and I'm trying to do easy wins before tearing up the sidewalk

mruberry replied:
Haha; OK.

mruberry commented on test/test_bundled_images.py:
```diff
@@ -2,10 +2,13 @@
-from torch.testing._internal.common_utils import TestCase
+from torch.testing._internal.common_utils import TestCase, IS_FBCODE
+
+if not IS_FBCODE:
```
Better add a comment here. What happens if this test file is run using unittest?

ezyang replied:
From OSS, it doesn't work. But that's true today anyway.

mruberry replied:
If this adds a comment explaining the pytest skip maybe a comment that this test won't work in OSS would be friendly, too? I guess there's a TODO to make this file internal-only or skip it when run with unittest, too.

ezyang replied:
Dumping this part of the patch, this test needs to go.

mruberry reviewed (approved):
This may still want to add a comment.

ezyang committed:
Update on "Marginally improve pytest collection for top-level test files"


I'm trying to make `pytest test/*.py` work--right now, it fails during
test collection.  This removes a few of the easier to fix pytest
collection problems one way or another.  I have two remaining problems
which is that the default dtype is trashed on entry to test_torch.py and
test_cuda.py, I'll try to fix those in a follow up.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26918377](https://our.internmc.facebook.com/intern/diff/D26918377)

[ghstack-poisoned]

rgommers reviewed (approved):
LGTM modulo the one comment on unconditional `import pytest` that I'm not sure about.

rgommers commented on test/test_cpp_extensions_aot.py:
```diff
@@ -8,10 +8,14 @@
 import torch.backends.cudnn
 import torch.utils.cpp_extension
 
+import pytest
```
The `calculate_shards` part of `test/run_test.py` is clear as mud, but I think the tests are run. Probably `pytest` is installed on the relevant CI jobs though - it looks like this should be added to `USE_PYTEST_LIST`.

rgommers replied:
What I mean is, the import may be missing currently, `pytest` is still an optional dependency today. In an env with no pytest installed:
```
$ python test/test_cpp_extensions_aot.py 
...
Traceback (most recent call last):
  File "test/test_cpp_extensions_aot.py", line 11, in <module>
    import pytest
ModuleNotFoundError: No module named 'pytest'
```

Install `pytest`, try again:
```
$ python test/test_cpp_extensions_aot.py 
...
Traceback (most recent call last):
  File "test/test_cpp_extensions_aot.py", line 14, in <module>
    import torch_test_cpp_extension.cpp as cpp_extension
ModuleNotFoundError: No module named 'torch_test_cpp_extension'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test/test_cpp_extensions_aot.py", line 21, in <module>
    ) from e
RuntimeError: test_cpp_extensions_aot.py cannot be invoked directly. Run `python run_test.py -i test_cpp_extensions_aot_ninja` instead
```

So at least it may hide the informative error message.

ezyang replied:
Oh! I thought we had agreed we can make it mandatory haha. OK, I can probably guard this a bit more.

rgommers replied:
Yes we can! And will! It's just not done yet.

pmeier reviewed (commented):
Nothing to add to the previous reviews.
